core/src/main/scala/kafka/tools/StorageTool.scala

```diff
@@ -198,9 +192,41 @@ object StorageTool extends Logging {
     }
   }
 
+  private def buildMetaProps(
+    clusterIdStr: String,
+    config: KafkaConfig
+  ): MetaProperties = {
+    val effectiveClusterId = try {
+      Uuid.fromString(clusterIdStr)
+    } catch {
+      case e: Throwable => throw new TerseFailure(s"Cluster ID string $clusterIdStr " +
+        s"does not appear to be a valid UUID: ${e.getMessage}")
+    }
+
+    val processRoles = config.processRoles.toSet
+
+    val brokerId = if (processRoles.contains(BrokerRole)) {
+      require(config.brokerId >= 0,
+        s"The `broker` process role requires non-negative `broker.id`, but found ${config.brokerId}")
+      Some(config.brokerId)
+    } else {
+      None
+    }
+
+    val controllerId = if (processRoles.contains(ControllerRole)) {
+      require(config.brokerId >= 0,
```

Comment:
`require(config.controllerId >= 0,`

It doesn't fix any of the 3 failing tests (indicating a gap in testing -- understandable at this point) but does fix this error if you try to format with a controller-only config file:

`bin/kafka-storage.sh format -c config/kip500-controller-node.properties --cluster-id 6bd37820-6745-4790-ae98-620300e1f61b`
```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: The `controller` process role requires non-negative `controller.id`, but found 3000
	at kafka.tools.StorageTool$.buildMetaProps(StorageTool.scala:218)
```